File: tests/with_tacs/crm.py
"""
This file is part of the package FUNtoFEM for coupled aeroelastic simulation
and design optimization.

Copyright (C) 2015 Georgia Tech Research Corporation.
Additional copyright (C) 2015 Kevin Jacobson, Jan Kiviaho and Graeme Kennedy.
All rights reserved.

FUNtoFEM is licensed under the Apache License, Version 2.0 (the "License");
you may not use this software except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

"""
--------------------------------------------------------------------------------
CRM (common research model)
--------------------------------------------------------------------------------
The following script demonstrates a "half-loop" in aeroelastic analysis using
TransferScheme 
"""
import numpy as np
from mpi4py import MPI
from tacs import TACS, elements, constitutive
from funtofem import TransferScheme

tacs_comm = MPI.COMM_WORLD
"""
--------------------------------------------------------------------------------
Input aerodynamic surface mesh and forces
--------------------------------------------------------------------------------
"""
XF = np.loadtxt('funtofemforces.dat')
aero_X = XF[:,:3].flatten().astype(TransferScheme.dtype)
aero_loads = (251.8 * 251.8 / 2 * 0.3)*XF[:,3:].flatten().astype(TransferScheme.dtype) / float(tacs_comm.Get_size())
aero_nnodes = aero_X.shape[0]/3

"""
--------------------------------------------------------------------------------
Initialize TACS
--------------------------------------------------------------------------------
"""

# Load structural mesh from bdf file
struct_mesh = TACS.MeshLoader(tacs_comm)
struct_mesh.scanBDFFile("CRM_box_2nd.bdf")

# Set constitutive properties
rho = 2500.0 # density, kg/m^3
E = 70e9 # elastic modulus, Pa
nu = 0.3 # poisson's ratio
kcorr = 5.0 / 6.0 # shear correction factor
ys = 350e6 # yield stress, Pa
min_thickness = 0.001
max_thickness = 0.020
thickness = 0.005

# Loop over components, creating stiffness and element object for each
num_components = struct_mesh.getNumComponents()
for i in range(num_components):
    descriptor = struct_mesh.getElementDescript(i)
    stiff = constitutive.isoFSDT(rho, E, nu, kcorr, ys, thickness, i,
                                 min_thickness, max_thickness)
    element = None
    if descriptor in ["CQUAD", "CQUADR", "CQUAD4"]:
        element = elements.MITCShell(2, stiff, component_num=i)
    struct_mesh.setElement(i, element)

# Create tacs assembler object
tacs = struct_mesh.createTACS(6)
res = tacs.createVec()
ans = tacs.createVec()
mat = tacs.createFEMat()

# Create distributed node vector from TACS Assembler object and extract the
# nodes
struct_X_vec = tacs.createNodeVec()
tacs.getNodes(struct_X_vec)
struct_X = struct_X_vec.getArray()
struct_nnodes = len(struct_X)/3

# Create the preconditioner for the corresponding matrix
pc = TACS.Pc(mat)

# Set design variables (thicknesses of components)
x = np.loadtxt('sizing.dat', dtype=TACS.dtype)
tacs.setDesignVars(x)

# Assemble the Jacobian
alpha = 1.0
beta = 0.0
gamma = 0.0
tacs.assembleJacobian(alpha, beta, gamma, res, mat)
pc.factor()

"""
--------------------------------------------------------------------------------
Set up MELD and transfer loads
--------------------------------------------------------------------------------
"""
# Creating MELD
comm = MPI.COMM_WORLD
isymm = -1
num_nearest = 100
beta = 0.5
meld = TransferScheme.pyMELD(comm, comm, 0, comm, 0, isymm, num_nearest, beta)

# Set nodes into transfer scheme
meld.setStructNodes(struct_X) 
meld.setAeroNodes(aero_X) 

# Initialize transfer scheme
meld.initialize()

# Transfer loads
init_disps = np.zeros(struct_X.shape, dtype=TransferScheme.dtype)
aero_disps = np.zeros(3*aero_nnodes, dtype=TransferScheme.dtype)
struct_loads = np.zeros(3*struct_nnodes, dtype=TransferScheme.dtype)

meld.transferDisps(init_disps, aero_disps)
meld.transferLoads(aero_loads, struct_loads)
meld_load_mags = np.sqrt(np.sum(struct_loads.reshape((3,1,-1))**2, axis=0))

# collect on the master
total_sum = comm.reduce(np.sum(struct_loads))
nnodes = comm.reduce(struct_nnodes)

if comm.Get_rank()==0:
    print("MELD sum loads:  ", total_sum)
    print("MELD mean load:  ", total_sum / (nnodes*3))
   # print "MELD load stdev: ", np.std(global_load_mags)

"""
--------------------------------------------------------------------------------
Set up RBF scheme and transfer loads
--------------------------------------------------------------------------------
"""
# Creating RBF
rbf_type = TransferScheme.PY_THIN_PLATE_SPLINE
denom = 10
rbf = TransferScheme.pyRBF(comm, comm, 0, comm, 0, rbf_type, denom)

# Set nodes into transfer scheme
rbf.setStructNodes(struct_X) 
rbf.setAeroNodes(aero_X) 

# Initialize transfer scheme
rbf.initialize()

# Transfer loads
aero_disps_rbf = np.zeros(3*aero_nnodes, dtype=TransferScheme.dtype)
struct_loads_rbf = np.zeros(3*struct_nnodes, dtype=TransferScheme.dtype)

rbf.transferLoads(aero_loads, struct_loads_rbf)
rbf_load_mags = np.sqrt(np.sum(struct_loads_rbf.reshape((3,1,-1))**2, axis=0))
print("RBF sum loads:  ", np.sum(struct_loads_rbf))
print("RBF mean load:  ", np.mean(rbf_load_mags))
print("RBF load stdev: ", np.std(rbf_load_mags))

"""
--------------------------------------------------------------------------------
Compute displacements due to MELD's loads using TACS
--------------------------------------------------------------------------------
"""
# Set structural loads into residual array res. Since we're using shell
# elements, res specifies three force components and three moment components per
# node
res.zeroEntries()
res_array = res.getArray()
res_array[0::6] = struct_loads[0::3]
res_array[1::6] = struct_loads[1::3]
res_array[2::6] = struct_loads[2::3]
tacs.applyBCs(res)

# Solve Ku = F and set u into answer array
pc.applyFactor(res, ans) 
tacs.setVariables(ans)

"""
# Analogous to res, ans specifies three displacements and three rotations per
# node
ans_array = ans.getArray()
struct_disps = np.zeros(len(struct_X), dtype=TransferScheme.dtype)
struct_disps[0::3] = ans_array[0::6]
struct_disps[1::3] = ans_array[1::6]
struct_disps[2::3] = ans_array[2::6]

# We can use TACS built-in visualization capability to look at the loads that
# came out of the load transfer. We do so by overwriting the rotations in ans
# with those loads
ans_array[3::6] = struct_loads[0::3]
ans_array[4::6] = struct_loads[1::3]
ans_array[5::6] = struct_loads[2::3]
tacs.applyBCs(ans)
tacs.setVariables(ans)
"""

# Output information from TACS for visualization in Tecplot
flag = (TACS.ToFH5.NODES |
        TACS.ToFH5.DISPLACEMENTS |
        TACS.ToFH5.STRAINS |
        TACS.ToFH5.STRESSES |
        TACS.ToFH5.EXTRAS)
f5 = TACS.ToFH5(tacs, TACS.PY_SHELL, flag)
f5.writeToFile('crm_meld.f5')

"""
--------------------------------------------------------------------------------
Compute displacements due to the RBF scheme's loads using TACS
--------------------------------------------------------------------------------
"""
# Set structural loads into residual array res. Since we're using shell
# elements, res specifies three force components and three moment components per
# node
res.zeroEntries()
res_array = res.getArray()
res_array[0::6] = struct_loads_rbf[0::3]
res_array[1::6] = struct_loads_rbf[1::3]
res_array[2::6] = struct_loads_rbf[2::3]
tacs.applyBCs(res)

# Solve Ku = F and set u into answer array
pc.applyFactor(res, ans) 
tacs.setVariables(ans)

"""
# Analogous to res, ans specifies three displacements and three rotations per
# node
ans_array = ans.getArray()
struct_disps = np.zeros(len(struct_X), dtype=TransferScheme.dtype)
struct_disps[0::3] = ans_array[0::6]
struct_disps[1::3] = ans_array[1::6]
struct_disps[2::3] = ans_array[2::6]

# We can use TACS built-in visualization capability to look at the loads that
# came out of the load transfer. We do so by overwriting the rotations in ans
# with those loads
ans_array[3::6] = struct_loads[0::3]
ans_array[4::6] = struct_loads[1::3]
ans_array[5::6] = struct_loads[2::3]
tacs.applyBCs(ans)
tacs.setVariables(ans)
"""

# Output information from TACS for visualization in Tecplot
f5.writeToFile('crm_rbf.f5')

"""
--------------------------------------------------------------------------------
Transfer displacments using TransferScheme
--------------------------------------------------------------------------------
"""
"""
meld.transferDisps(struct_disps, aero_disps)
rbf.transferDisps(struct_disps, aero_disps_rbf)
"""

File: examples/framework/functionality/with_tacs/crm.c
/*
  This file is part of the package FUNtoFEM for coupled aeroelastic simulation
  and design optimization.

  Copyright (C) 2015 Georgia Tech Research Corporation.
  Additional copyright (C) 2015 Kevin Jacobson, Jan Kiviaho and Graeme Kennedy.
  All rights reserved.

  FUNtoFEM is licensed under the Apache License, Version 2.0 (the "License");
  you may not use this software except in compliance with the License.
  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/
#include "TACSMeshLoader.h"
#include "MITCShell.h"
#include "isoFSDTStiffness.h"
#include "TransferScheme.h"
#include "MELD.h"

int main( int argc, char *argv[] ){
  // Initialize MPI and declare communicator
  MPI_Init(&argc, &argv);
  MPI_Comm comm = MPI_COMM_WORLD;

/*   
  In TACS
  -----------
  Load in the structural mesh, create TACS from the structural mesh, assemble
  the stiffness matrix, and load in the aerodynamic surface mesh
*/

  // Load structural mesh from BDF file using TACS MeshLoader
  const char *filename = "../resources/CRM_box_2nd.bdf";
  TACSMeshLoader *mesh = new TACSMeshLoader(comm);
  mesh->incref();
  mesh->scanBDFFile(filename);

  // Get number of components specified in BDF file
  int num_components = mesh->getNumComponents();

  // Set properties needed to create constitutive object 
  double rho = 2500.0; // density, kg/m^3
  double E = 70e9; // elastic modulus, Pa
  double nu = 0.3; // poisson's ratio
  double kcorr = 5.0/6.0; // shear correction factor
  double ys = 350e6; // yield stress, Pa

  // Loop over components, creating constituitive object for each
  for ( int i = 0; i < num_components; i++ ){
    const char *descriptor = mesh->getElementDescript(i);
    double min_thickness = 0.01;
    double max_thickness = 0.20;
    double thickness = 0.07;
    isoFSDTStiffness *stiff = new isoFSDTStiffness(rho, E, nu, kcorr, ys,
        thickness, i, min_thickness, max_thickness); 

    // Initialize element object
    TACSElement *element = NULL;

    // Create element object using constituitive information and type defined in
    // descriptor
    if ( strcmp(descriptor, "CQUAD") == 0 || 
         strcmp(descriptor, "CQUADR") == 0 ||
         strcmp(descriptor, "CQUAD4") == 0) {
      element = new MITCShell<2>(stiff, LINEAR, i);
    }
    mesh->setElement(i, element);
  }

  // Create TACS Assembler from TACS MeshLoader
  TACSAssembler *tacs = mesh->createTACS(6);
  tacs->incref();
  mesh->decref();

  // Create matrix and vectors 
  TACSBVec *ans = tacs->createVec(); // displacements and rotations
  TACSBVec *f = tacs->createVec(); // loads
  FEMat *mat = tacs->createFEMat(); // preconditioner
  TACSBVec *struct_X_vec = tacs->createNodeVec();

  // Increment reference count to the matrix/vectors
  ans->incref();
  f->incref();
  mat->incref();
  struct_X_vec->incref();

  // Allocate the factorization
  int lev = 10000;
  double fill = 10.0;
  int reorder_schur = 1;
  PcScMat *pc = new PcScMat(mat, lev, fill, reorder_schur); 
  pc->incref();

  // Assemble and factor the stiffness matrix
  double alpha = 1.0, beta = 0.0, gamma = 0.0;
  tacs->assembleJacobian(alpha, beta, gamma, NULL, mat);
  tacs->applyBCs(mat);
  pc->factor();

  // Retrieve node locations and number of nodes for structual mesh
  int struct_nnodes = tacs->getNumNodes();
  tacs->getNodes(struct_X_vec);
  TacsScalar *struct_X;
  int struct_size = struct_X_vec->getArray(&struct_X);

  // Load aerodynamic mesh using TACS MeshLoader
  const char *aero_filename = "../resources/ucrm_aero_mesh.bdf";
  TACSMeshLoader *aero_mesh = new TACSMeshLoader(comm);
  aero_mesh->incref();
  aero_mesh->scanBDFFile(aero_filename);

  // Get number of components prescribed in BDF file
  int aero_num_components = aero_mesh->getNumComponents();

  // Loop over components, creating constituitive object for each
  for ( int i = 0; i < aero_num_components; i++ ){
    const char *aero_descriptor = aero_mesh->getElementDescript(i);
    double min_thickness = 0.01;
    double max_thickness = 0.20;
    double thickness = 0.07;
    isoFSDTStiffness *aero_stiff = new isoFSDTStiffness(rho, E, nu, kcorr, ys,
        thickness, i, min_thickness, max_thickness); 

    // Initialize element object
    TACSElement *element = NULL;

    // Create element object using constituitive information and type defined in
    // descriptor
    if ( strcmp(aero_descriptor, "CQUAD") == 0 || 
         strcmp(aero_descriptor, "CQUADR") == 0 ||
         strcmp(aero_descriptor, "CQUAD4") == 0) {
      element = new MITCShell<2>(aero_stiff, LINEAR, i);
    }
    aero_mesh->setElement(i, element);
  }

  // Create tacs assembler from mesh loader object
  TACSAssembler *aero_tacs = aero_mesh->createTACS(6);
  aero_tacs->incref();
  aero_mesh->decref();

  // Retrieve node locations and number of nodes for aero mesh from TACS
  TACSBVec *aero_X_vec = aero_tacs->createNodeVec();
  aero_X_vec->incref();
  int aero_nnodes = aero_tacs->getNumNodes();
  aero_tacs->getNodes(aero_X_vec);
  TacsScalar *aero_X;
  int aero_size = aero_X_vec->getArray(&aero_X);

/*   
  In TransferScheme
  -----------------
  Loading in the meshes, inventing loads, and transferring loads
*/

  // Specify scheme type, symmetry, number of nearest nodes
  int isymm = -1;
  int num_nearest = 100;
  F2FScalar decay_param = 0.5;

  // Create instance of transfer scheme class
  TransferScheme *meld = new MELD(comm, comm, 0, comm, 0, isymm, num_nearest,
                                  decay_param);

  // Initialize transfer scheme
  meld->setStructNodes(struct_X, struct_nnodes);
  meld->setAeroNodes(aero_X, aero_nnodes);
  meld->initialize();

  // Create loads on aero mesh and apply load transfer
  F2FScalar *aero_loads = new F2FScalar[ 3*aero_nnodes ];
  for ( int i = 0 ; i < aero_nnodes; i++ ) {
    aero_loads[3*i+0] = 0.0;
    aero_loads[3*i+1] = 0.0;
    aero_loads[3*i+2] = 10000.0;
  }
  F2FScalar *init_disps = new F2FScalar[3*struct_nnodes];
  memset(init_disps, 0.0, 3*struct_nnodes*sizeof(F2FScalar)); 
  F2FScalar *aero_disps = new F2FScalar[3*aero_nnodes];
  memset(aero_disps, 0.0, 3*aero_nnodes*sizeof(F2FScalar)); 
  F2FScalar *struct_loads = new F2FScalar[3*struct_nnodes];
  memset(struct_loads, 0.0, 3*struct_nnodes*sizeof(F2FScalar)); 

  meld->transferDisps(init_disps, aero_disps);
  meld->transferLoads(aero_loads, struct_loads);

/*   
  In TACS
  -----------
  Setting the loads, solving, retrieving displacements, configuring output
*/

  // Set load vector in TACS to load vector from transfer scheme
  // (For shell elements, first 3 degrees of freedom in residual vector f
  // correspond to forces; last 3 correspond to moments. We will populate the
  // first 3.)
  TacsScalar *f_array;
  f->zeroEntries();
  int f_size = f->getArray(&f_array);
  for ( int i = 0; i < struct_nnodes; i++ ){
    memcpy(&f_array[6*i], &struct_loads[3*i], 3*sizeof(TacsScalar));
  }
  tacs->applyBCs(f);

  // Solve and set solution into ans vector
  pc->applyFactor(f, ans);
  tacs->setVariables(ans);
  
  // Extract displacements from the ans vector
  // (For shell elements, first 3 degrees of freedom correspond to u,v,w
  // displacements; last 3 correspond to rotations)
  TacsScalar *struct_disps = new TacsScalar[ 3*struct_nnodes ];
  memset(struct_disps, 0, 3*struct_nnodes*sizeof(TacsScalar));
  TacsScalar *ans_array;
  ans->getArray(&ans_array);
  for ( int i = 0; i < struct_nnodes; i++ ){
    memcpy(&struct_disps[3*i], &ans_array[6*i], 3*sizeof(TacsScalar));
  }

  // Overwrite rotations in ans vector with forces computing by transfer scheme in
  // order to visualize them in TACS output
  for ( int i = 0; i < struct_nnodes; i++ ){
    memcpy(&ans_array[6*i+3], &f_array[3*i], 3*sizeof(TacsScalar));
  }
  tacs->setVariables(ans);

  // Create an TACSToFH5 object for writing output to files
  unsigned int write_flag = (TACSElement::OUTPUT_NODES |
                             TACSElement::OUTPUT_DISPLACEMENTS |
                             TACSElement::OUTPUT_STRAINS |
                             TACSElement::OUTPUT_STRESSES |
                             TACSElement::OUTPUT_EXTRAS);
  TACSToFH5 * f5 = new TACSToFH5(tacs, TACS_SHELL, write_flag);
  f5->incref();

  // Write the displacements
  f5->writeToFile("ucrm.f5");

/*   
  In transfer scheme
  -----------
  Transfer the displacements computed by TACS and output data file
*/

  // Transfer displacements
  meld->transferDisps(struct_disps, aero_disps);

/*   
  Clean up
  ---------
*/

  // Free all the allocated data
  delete [] init_disps;
  delete [] aero_disps;
  delete [] aero_loads;
  delete [] struct_loads;

  // Call transfer scheme destructor
  delete meld;

  // Decrease the reference count for TACS objects
  pc->decref();
  mat->decref();
  ans->decref();
  f->decref();
  struct_X_vec->decref();
  aero_X_vec->decref();
  f5->decref();
  tacs->decref();
  aero_tacs->decref();

  // Finalize MPI and return
  MPI_Finalize();
  return (0);
}
